### PR TITLE
Request to add MM-JOKER motherboard，Thanks

### PR DESCRIPTION
**Pull request by magicmaker3**
### Description

Hello, this motherboard is integrated with the motherboard and control screen, and can also be disconnected using cable connections. The firmware is modified based on MRR_ESPE, and has a 4-axis control and OLED screen. It retains necessary basic functions and has a good experience. It also has 3 native IO and 595 interfaces to support expansion, and has designed expansion modules for 2-axis and 5-axis.

### Requirements

The motherboard and screen are integrated, only supporting 4-wire 12864 OLED, with driver 1306 or 1315

### Benefits

experience is good, the scalability is strong, the PCB layout is reasonable, the design is beautiful, the cost is low, and the price is low. Over 1500 pieces have been sold.

### Configurations

- Modified `marlin/src/pins/pins.h,marlin/src/core/boards.h`
- Added `marlin/src/pins/esp32/pins_MM_JOKER.h`

### Related Issues

ESP32 has two IOs that pull up or down, which can cause it to not start up and can only be used for specific purposes. I have changed these two IOs to encoders, making the originally poor IOs slightly more abundant. In addition to meeting most basic functions, it also supports material breakage detection and BLTouch, and has also released three customizable IOs for expansion and distribution.

### CODE REVIEW

File: Marlin/src/pins/esp32/pins_MM_JOKER.h [hunk 1]
```diff
@@ -0,0 +1,258 @@
+/**
+ * JIOKER
```
Comment (GMagician) [inline Marlin/src/pins/esp32/pins_MM_JOKER.h]: typo error?

**Comment by thinkyhead** [comment]
Haha, when I saw no text in the description I had a hunch you had put your text in between the hidden comment delimiters `<!-- -->`. Fixed!

File: Marlin/src/pins/esp32/pins_MM_JOKER.h [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+ *
+ */
+#pragma once
+
+/**
+ * MRR ESPE pin assignments
+ *
+ * 3D printer control board based on the ESP32 microcontroller.
+ * Supports 5 stepper drivers (using I2S stepper stream), heated bed,
+ * single hotend, and LCD controller.
```
Comment (thinkyhead) [inline Marlin/src/pins/esp32/pins_MM_JOKER.h]: Please update this description if you have any information about this board.

File: Marlin/src/pins/esp32/pins_MM_JOKER.h [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+#include "env_validate.h"
+
+//#if EXTRUDERS > 2 || E_STEPPERS > 2
+  //#error "MRR ESPE supports up to 2 E steppers."
+//#elif HAS_MULTI_HOTEND
+  //#error "MRR ESPE only supports 1 hotend / E stepper."
+//#endif
+
+#define BOARD_INFO_NAME      "MRR ESPE"
+#define BOARD_WEBSITE_URL    "github.com/maplerainresearch/MRR_ESPE"
```
Comment (thinkyhead) [inline Marlin/src/pins/esp32/pins_MM_JOKER.h]: Please update the board name and link.

File: Marlin/src/pins/esp32/pins_MM_JOKER.h [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+//
+//拓展板
+//
+//+2只有2个MAX限位，+5的限位和热敏只能6选3，开启需注释掉46行Limit Switches
+#ifndef X_STOP_PIN
+  #ifndef X_MIN_PIN
+    #define X_MIN_PIN                          15
+  #endif
+  #ifndef X_MAX_PIN
+    #define X_MAX_PIN                          -1//32
```
Comment (thinkyhead) [inline Marlin/src/pins/esp32/pins_MM_JOKER.h]: Is there any situation with this board where pins 32, 33, 34 are expected to be used as endstops?

**Comment by thinkyhead** [comment]
@magicmaker3 — Are you still with us here on Earth? There are some questions about this new board.

**Comment by magicmaker3** [comment]
Sorry, I only saw the email today. May I ask what the problem is

**Comment by magicmaker3** [comment]
Yes, sorry, it should be JOKER

**Comment by magicmaker3** [comment]
> Please update the board name and link.

Okay, I will update this section as soon as possible

**Comment by magicmaker3** [comment]
Okay, I will update this section as soon as possible



At 2023-05-30 06:53:41, "Scott Lahteine" ***@***.***> wrote:

@thinkyhead commented on this pull request.

In Marlin/src/pins/esp32/pins_MM_JOKER.h:

> + *
+ * 3D printer control board based on the ESP32 microcontroller.
+ * Supports 5 stepper drivers (using I2S stepper stream), heated bed,
+ * single hotend, and LCD controller.
+ */
+
+#include "env_validate.h"
+
+//#if EXTRUDERS > 2 || E_STEPPERS > 2
+  //#error "MRR ESPE supports up to 2 E steppers."
+//#elif HAS_MULTI_HOTEND
+  //#error "MRR ESPE only supports 1 hotend / E stepper."
+//#endif
+
+#define BOARD_INFO_NAME      "MRR ESPE"
+#define BOARD_WEBSITE_URL    "github.com/maplerainresearch/MRR_ESPE"


Please update the board name and link.

**Comment by magicmaker3** [comment]
> Is there any situation with this board where pins 32, 33, 34 are expected to be used as endstops?

Please refer to the pictures below. I will keep these IO for expansion or customization

**Comment by magicmaker3** [comment]
Sorry, I have been waiting for your message on GitHub for the past few days, but I didn't expect it to be through email. I only saw it today when I logged in to my email. Thank you for your response. I have answered all the questions one by one, and then I will update the ones that need to be updated

**Comment by magicmaker3** [comment]
Sorry, it's the first time I've done this

**Comment by thinkyhead** [comment]
> Please refer to the pictures below. I will keep these IO for expansion or customization

We can't see things that you attach to email replies. Please come over to the PR page and post your answers here. https://github.com/MarlinFirmware/Marlin/pull/25897

**Comment by magicmaker3** [comment]
JOKER-PIN口图
joker+2
joker+5
joker2 0
JOKER+2-PIN口图
JOKER+5-PIN口图

**Comment by magicmaker3** [comment]
Hello, I have sent the pictures up

**Comment by thinkyhead** [comment]
> Hello, I have sent the pictures up

Ah yes. I could tell by the presence of the pictures. Thank you for posting them!

**Comment by magicmaker3** [comment]
Do you need any other information, such as schematic  schematic circuit

**Pull request merged**